README
README
======

This is unofficial port of netsyslog library written for python2 to python3.
Library api should be fully compatible with original netsyslog library.

For more information visit pages of original project http://hacksaw.sourceforge.net/netsyslog/


CHANGELOG.rst
Changelog
=========

0.1.1
-----
    - Changed packaging library.

0.1.0
-----
    - Ported to python3.


run_tests.sh
#!/usr/bin/env sh

py.test tests/ $@


tests/netsyslog_test.py
#!/usr/bin/env python3
import os
import socket
import syslog
import time
import mock
import pytest
import sys

import netsyslog

DEFAULT_TIMESTAMP = "Jun  7 09:00:00"
DEFAULT_HOSTNAME = "myhost"
DEFAULT_HEADER = "%s %s" % (DEFAULT_TIMESTAMP, DEFAULT_HOSTNAME)


@pytest.fixture(scope='function')
def sendto():
    orig = socket.socket.sendto
    mock_socket = mock.Mock()
    socket.socket.sendto = mock_socket
    yield mock_socket
    socket.socket.sendto = orig


@pytest.fixture(scope='function')
def localtime():
    orig = time.localtime
    time.localtime = mock_local_time
    yield mock_local_time
    time.localtime = orig


@pytest.fixture(scope='function')
def gethostname():
    orig = socket.gethostname
    socket.gethostname = mock_hostname
    yield mock_hostname
    socket.gethostname = orig


def mock_local_time():
    return 2005, 6, 7, 9, 0, 0, 1, 158, 1


def mock_hostname():
    return DEFAULT_HOSTNAME


def test_priority_format():
    """Check PRI is correctly formatted"""
    pri = netsyslog.PriPart(syslog.LOG_LOCAL4, syslog.LOG_NOTICE)
    assert str(pri) == "<165>"


def test_automatic_timestamp(localtime, gethostname):
    """Check HEADER is automatically calculated if not set"""
    header = netsyslog.HeaderPart()
    assert str(header) == " ".join((DEFAULT_TIMESTAMP, DEFAULT_HOSTNAME))


def test_incorrect_characters_disallowed(localtime, gethostname):
    """Check only valid characters are used in the HEADER"""
    # Only allowed characters are ABNF VCHAR values and space.
    # Basically, if ord() returns between 32 and 126 inclusive
    # it's okay.
    bad_char = "\x1f"  # printable, ord() returns 31
    header = netsyslog.HeaderPart()
    header.timestamp = header.timestamp[:-1] + bad_char
    assert str(header) == " ".join((DEFAULT_TIMESTAMP, DEFAULT_HOSTNAME))


def test_set_timestamp_manually(localtime, gethostname):
    """Check it is possible to set the timestamp in HEADER manually"""
    timestamp = "Jan 31 18:12:34"
    header = netsyslog.HeaderPart(timestamp=timestamp)
    assert str(header) == "%s %s" % (timestamp, DEFAULT_HOSTNAME)


def test_set_hostname_manually(localtime, gethostname):
    """Check it is possible to set the hostname in HEADER manually"""
    hostname = "otherhost"
    header = netsyslog.HeaderPart(hostname=hostname)
    assert str(header) == "%s %s" % (DEFAULT_TIMESTAMP, hostname)


# check format of time and hostname, set automatically if incorrect
#   - time is "Mmm dd hh:mm:ss" where dd has leading space, hh leading 0
#   - single space between time and hostname
#   - no space in hostname
#   - if using hostname, not IP, no dots allowed
# print message to stderr if badly formatted message encountered


DEFAULT_TAG = "program"
MOCK_PID = 1234


@pytest.fixture(scope='function')
def sys_args():
    orig = sys.argv
    sys.argv = [DEFAULT_TAG]
    yield sys.argv
    sys.argv = orig


@pytest.fixture(scope='function')
def get_pid():

    def mock_pid():
        return MOCK_PID

    orig = os.getpid
    os.getpid = mock_pid
    yield mock_pid
    os.getpid = orig


def test_tag_defaults_to_progname(sys_args):
    """Check TAG defaults to program name"""
    msg = netsyslog.MsgPart()
    assert msg.tag == DEFAULT_TAG


def test_override_tag():
    """Check TAG can be set manually"""
    msg = netsyslog.MsgPart(tag="mytag")
    assert msg.tag == "mytag"


def test_tag_trimmed_if_too_long():
    """Check long TAGs are trimmed to 32 characters"""
    tag = "abcd" * 10
    msg = netsyslog.MsgPart(tag=tag)
    assert msg.tag == tag[:32]


def test_space_prefixed_to_content():
    """Check single space inserted infront of CONTENT if necessary"""
    msg = netsyslog.MsgPart("program", content="hello")
    assert str(msg) == "program: hello"


def test_space_only_added_if_necessary():
    """Check space only added to CONTENT if necessary"""
    msg = netsyslog.MsgPart("program", content=" hello")
    assert str(msg) == "program hello"


def test_include_pid():
    """Check the program's pid can be included in CONTENT"""
    msg = netsyslog.MsgPart("program", "hello", pid=MOCK_PID)
    assert str(msg) == "program[%d]: hello" % MOCK_PID


DEFAULT_PRI = netsyslog.PriPart(syslog.LOG_LOCAL4, syslog.LOG_NOTICE)
DEFAULT_HEADER = netsyslog.HeaderPart(DEFAULT_TIMESTAMP, DEFAULT_HOSTNAME)
DEFAULT_MSG = netsyslog.MsgPart(DEFAULT_TAG, "hello")


def test_message_format():
    """Check syslog message is correctly constructed"""
    packet = netsyslog.Packet(DEFAULT_PRI, DEFAULT_HEADER, DEFAULT_MSG)
    header = " ".join((DEFAULT_TIMESTAMP, DEFAULT_HOSTNAME))
    start_of_packet = "<165>%s %s" % (header, DEFAULT_TAG)
    assert str(packet).startswith(start_of_packet)


def test_max_length():
    """Check that no syslog packet is longer than 1024 bytes"""
    message = "a" * 2048
    packet = netsyslog.Packet(DEFAULT_PRI, DEFAULT_HEADER, message)
    assert len(str(packet)) == netsyslog.Packet.MAX_LEN


def test_send_message(sendto, gethostname, localtime, sys_args):
    """Check we can send a message via UDP"""
    logger = netsyslog.Logger()

    packet = netsyslog.Packet(DEFAULT_PRI, DEFAULT_HEADER, DEFAULT_MSG)

    hostname = "localhost"
    address_1 = (hostname, netsyslog.Logger.PORT)
    logger.add_host(hostname)

    logger.log(syslog.LOG_LOCAL4, syslog.LOG_NOTICE, "hello")
    sendto.assert_called_once_with(
        bytes(str(packet), encoding='ascii'), address_1)


def test_send_message_to_multiple(sendto, gethostname, localtime, sys_args):
    logger = netsyslog.Logger()

    packet = netsyslog.Packet(DEFAULT_PRI, DEFAULT_HEADER, DEFAULT_MSG)

    hostname = "localhost"
    address_1 = (hostname, netsyslog.Logger.PORT)
    logger.add_host(hostname)

    hostname = "remotehost"
    address_2 = (hostname, netsyslog.Logger.PORT)
    logger.add_host(hostname)

    logger.log(syslog.LOG_LOCAL4, syslog.LOG_NOTICE, "hello")

    assert sendto.call_count == 2
    sendto.assert_has_calls(
        (
            mock.call(bytes(str(packet), encoding='ascii'), address_1),
            mock.call(bytes(str(packet), encoding='ascii'), address_2)
        ),
        any_order=True
    )


def test_remove_host(sendto, gethostname, localtime, sys_args):
    """Check host can be removed from list of those receiving messages"""
    hostname = "localhost"
    logger = netsyslog.Logger()
    logger.add_host(hostname)
    logger.remove_host(hostname)
    logger.log(syslog.LOG_LOCAL4, syslog.LOG_NOTICE, "hello")
    assert sendto.call_count == 0


def test_pid_not_included_by_default(sendto, gethostname, localtime, sys_args):
    """Check the program's pid is not included by default"""
    packet = "<165>%s myhost program: hello" % DEFAULT_TIMESTAMP
    hostname = "localhost"
    address = (hostname, netsyslog.Logger.PORT)

    logger = netsyslog.Logger()
    logger.add_host(hostname)
    logger.log(syslog.LOG_LOCAL4, syslog.LOG_NOTICE, "hello")
    sendto.assert_called_once_with(bytes(packet, encoding='ascii'), address)


def test_include_pid_logger_test(sendto, gethostname, localtime, sys_args, get_pid):
    """Check the program's pid can be included in a log message"""
    packet = "<165>%s myhost program[1234]: hello" % DEFAULT_TIMESTAMP
    hostname = "localhost"
    address = (hostname, netsyslog.Logger.PORT)

    logger = netsyslog.Logger()
    logger.add_host(hostname)
    logger.log(syslog.LOG_LOCAL4, syslog.LOG_NOTICE, "hello", pid=True)

    sendto.assert_called_once_with(bytes(packet, encoding='ascii'), address)


def test_send_packets_by_hand(sendto):
    """Check we can send a hand crafted log packet"""
    hostname = "localhost"
    address = (hostname, netsyslog.Logger.PORT)
    packet_text = "<165>Jan  1 10:00:00 myhost myprog: hello"

    pri = netsyslog.PriPart(syslog.LOG_LOCAL4, syslog.LOG_NOTICE)
    header = netsyslog.HeaderPart("Jan  1 10:00:00", "myhost")
    msg = netsyslog.MsgPart("myprog", "hello")
    packet = netsyslog.Packet(pri, header, msg)
    logger = netsyslog.Logger()
    logger.add_host(hostname)
    logger.send_packet(packet)
    sendto.assert_called_once_with(
        bytes(packet_text, encoding='ascii'),
        address)


netsyslog.py
# Copyright (C) 2005 Graham Ashton <ashtong@users.sourceforge.net>
#
# This module is free software, and you may redistribute it and/or modify
# it under the same terms as Python itself, so long as this copyright message
# and disclaimer are retained in their original form.
#
# IN NO EVENT SHALL THE AUTHOR BE LIABLE TO ANY PARTY FOR DIRECT, INDIRECT,
# SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OF
# THIS CODE, EVEN IF THE AUTHOR HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
# DAMAGE.
#
# THE AUTHOR SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
# PARTICULAR PURPOSE.  THE CODE PROVIDED HEREUNDER IS ON AN "AS IS" BASIS,
# AND THERE IS NO OBLIGATION WHATSOEVER TO PROVIDE MAINTENANCE,
# SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
#
# $Id: netsyslog.py,v 1.9 2005/11/22 16:35:40 ashtong Exp $


"""netsyslog enables you to construct syslog messages and send them
(via UDP) to a remote syslog server directly from Python. You can send
log messages that contain the current time, local hostname and calling
program name (i.e. the typical requirement of a logging package) to
one or more syslog servers.

Unlike other syslog modules netsyslog also allows you to set the
metadata (e.g. time, host name, program name, etc.) yourself, giving
you full control over the contents of the UDP packets that it creates.

See L{Logger.log} and L{Logger.send_packet} for a synopsis of these
two techniques.

The format of the UDP packets sent by netsyslog adheres closely to
that defined in U{RFC 3164<http://www.ietf.org/rfc/rfc3164.txt>}. Much
of the terminology used in the RFC has been incorporated into the
names of the classes and properties and is used throughout this
documentation.

Further information and support can be found from the U{netsyslog home
page<http://hacksaw.sourceforge.net/netsyslog/>}.

"""


__version__ = "0.1.1"


import os
import socket
import sys
import time


class PriPart:

    """The PRI part of the packet.

    Though never printed in the output from a syslog server, the PRI
    part is a crucial part of the packet. It encodes both the facility
    and severity of the packet, both of which are defined in terms of
    numerical constants from the standard syslog module.

    See Section 4.1.1 of RFC 3164 for details.

    """

    def __init__(self, facility, severity):
        """Initialise the object, specifying facility and severity.

        Specify the arguments using constants from the syslog module
        (e.g. syslog.LOG_USER, syslog.LOG_INFO).

        """
        assert facility is not None
        assert severity is not None
        self.facility = facility
        self.severity = severity

    def __str__(self):
        value = self.facility + self.severity
        return "<%s>" % value


class HeaderPart:

    """The HEADER part of the message.

    The HEADER contains a timestamp and a hostname. It is the first
    component of the log message that is displayed in the output of
    syslog.

    See Section 4.1.2 of RFC 3164 for details.

    """

    def __init__(self, timestamp=None, hostname=None):
        """Initialise the object, specifying timestamp and hostname.

        The timestamp represents the local time when the log message
        was created. If the timestamp is not set the current local
        time will be used. See the L{HeaderPart.timestamp} property
        for a note on the format.

        The hostname should be set to the hostname of the computer
        that originally generated the log message. If the hostname is
        not set the hostname of the local computer will be used. See
        the L{HeaderPart.hostname} property for a note on the format.

        """
        self.timestamp = timestamp
        self.hostname = hostname

    def __str__(self):
        return "%s %s" % (self.timestamp, self.hostname)

    def _get_timestamp(self):
        return self._timestamp

    def _calculate_current_timestamp(self):
        localtime = time.localtime()
        day = time.strftime("%d", localtime)
        if day[0] == "0":
            day = " " + day[1:]
        value = time.strftime("%b %%s %H:%M:%S", localtime)
        return value % day

    def _timestamp_is_valid(self, value):
        if value is None:
            return False
        for char in value:
            if ord(char) < 32 or ord(char) > 126:
                return False
        return True
    
    def _set_timestamp(self, value):
        if not self._timestamp_is_valid(value):
            value = self._calculate_current_timestamp()
        self._timestamp = value

    timestamp = property(_get_timestamp, _set_timestamp, None,
                         """The local time when the message was written.

                         Must follow the format 'Mmm DD HH:MM:SS'.  If
                         the day of the month is less than 10, then it
                         MUST be represented as a space and then the
                         number.

                         """)

    def _get_hostname(self):
        return self._hostname

    def _set_hostname(self, value):
        if value is None:
            value = socket.gethostname()
        self._hostname = value

    hostname = property(_get_hostname, _set_hostname, None,
                        """The hostname where the log message was created.

                        Should be the first part of the hostname, or
                        an IP address. Should NOT be set to a fully
                        qualified domain name.

                        """)


class MsgPart:

    """Represents the MSG part of a syslog packet.

    The MSG part of the packet consists of the TAG and CONTENT. The
    TAG and the CONTENT fields must be separated by a non-alphanumeric
    character. Unless you ensure that the CONTENT field begins with
    such a character a seperator of a colon and space will be inserted
    between them when the C{MsgPart} object is converted into a UDP
    packet.

    See Section 4.1.3 of RFC 3164 for details.

    """

    MAX_TAG_LEN = 32

    def __init__(self, tag=None, content="", pid=None):
        """Initialise the object, specifying tag and content.

        See the documentation for the L{MsgPart.tag} and
        L{MsgPart.content} properties for further documentation.

        If the pid is set it will be prepended to the content in
        square brackets when the packet is created.

        """        
        self.tag = tag
        self.content = content
        self.pid = pid

    def __str__(self):
        content = self.content
        if self.pid is not None:
            content = "[%s]" % self.pid + content
        return self.tag + content

    def _get_tag(self):
        return self._tag

    def _set_tag(self, value):
        if value is None:
            value = sys.argv[0]
        self._tag = value[:self.MAX_TAG_LEN]

    tag = property(_get_tag, _set_tag, None,
                   """The name of the program that generated the log message.

                   The tag can only contain alphanumeric
                   characters. If the tag is longer than %d characters
                   it will be truncated automatically.

                   """ % MAX_TAG_LEN)

    def _get_content(self):
        return self._content

    def _prepend_seperator(self, value):
        try:
            first_char = value[0]
        except IndexError:
            pass
        else:
            if first_char.isalnum():
                value = ": " + value
        return value

    def _set_content(self, value):
        value = self._prepend_seperator(value)
        self._content = value

    content = property(_get_content, _set_content, None,
                       """The main component of the log message.

                       The content field is a freeform field that
                       often begins with the process ID (pid) of the
                       program that created the message.

                       """)


class Packet:

    """Combines the PRI, HEADER and MSG into a packet.

    If the packet is longer than L{MAX_LEN} bytes in length it is
    automatically truncated prior to sending; any extraneous bytes are
    lost.

    """

    MAX_LEN = 1024

    def __init__(self, pri, header, msg):
        """Initialise the object.

        The three arguments must be instances of the L{PriPart},
        L{HeaderPart} and L{MsgPart} classes.

        """
        self.pri = pri
        self.header = header
        self.msg = msg

    def __str__(self):
        message = "%s%s %s" % (self.pri, self.header, self.msg)
        return message[:self.MAX_LEN]


class Logger:

    """Send log messages to syslog servers.

    The Logger class provides two different methods for sending log
    messages. The first approach (the L{log} method) is suitable for
    creating new log messages from within a normal application. The
    second (the L{send_packet} method) is designed for use in
    circumstances where you need full control over the contents of
    the syslog packet.

    """

    PORT = 514

    def __init__(self):
        self._sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self._hostnames = {}

    def add_host(self, hostname):
        """Add hostname to the list of hosts that will receive packets.

        Can be a hostname or an IP address. Note that if the hostname
        cannot be resolved calls to L{log} or L{send_packet} will take
        a long time to return.
        
        """
        self._hostnames[hostname] = 1

    def remove_host(self, hostname):
        """Remove hostname from the list of hosts that will receive packets."""
        del self._hostnames[hostname]

    def _send_packet_to_hosts(self, packet):
        for hostname in self._hostnames:
            self._sock.sendto(bytes(str(packet), encoding='ascii'), (hostname, self.PORT))

    def log(self, facility, level, text, pid=False):
        """Send the message text to all registered hosts.

        The facility and level will be used to create the packet's PRI
        part. The HEADER will be automatically determined from the
        current time and hostname. The MSG will be set from the
        running program's name and the text parameter.

        This is the simplest way to use netsyslog, creating log
        messages containing the current time, hostname, program name,
        etc. This is how you do it::
        
            logger = netsyslog.Logger()
            logger.add_host("localhost")
            logger.log(syslog.LOG_USER, syslog.LOG_INFO, "Hello World")

        If pid is True the process ID will be prepended to the text
        parameter, enclosed in square brackets and followed by a
        colon.

        """
        pri = PriPart(facility, level)
        header = HeaderPart()
        if pid:
            msg = MsgPart(content=text, pid=os.getpid())
        else:
            msg = MsgPart(content=text)
        packet = Packet(pri, header, msg)
        self._send_packet_to_hosts(packet)

    def send_packet(self, packet):
        """Send a L{Packet} object to all registered hosts.

        This method requires more effort than L{log} as you need to
        construct your own L{Packet} object beforehand, but it does
        give you full control over the contents of the packet::

            pri = netsyslog.PriPart(syslog.LOG_USER, syslog.LOG_INFO)
            header = netsyslog.HeaderPart("Jun  1 18:34:03", "myhost")
            msg = netsyslog.MsgPart("myprog", "Hello World", mypid)
            packet = netsyslog.Packet(pri, header, msg)

            logger = netsyslog.Logger()
            logger.add_host("localhost")
            logger.send_packet(packet)

        """
        self._send_packet_to_hosts(packet)
